Привет! Меня зовут Исмаил, я буду проверять твой проект :) Можешь обращаться ко мне на «ты». Если тебе комфортно, то и я буду к тебе так обращаться, если нет, то обязательно скажи об этом.

Пожалуйста, не удаляй мои комментарии, которые я буду оставлять в работе. Это особенно поможет, если твои проекты будут отправлены на повторную проверку. При повторных проверках у комментариев будут приписки: "Вторая итерация", "Третья итерация" и т.д. 

Ты тоже можешь реагировать на мои комментарии, но в таком случае постарайся, чтобы твои комментарии отличались от моих: например, выделяй их своим любимым цветом — так у нас не возникнет путаницы :)

Кстати, про цвета! Мои комментарии будут в следующей цветовой гамме: зелеными, желтыми и красными. Например:

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
В случае, если решение на отдельном шаге является полностью правильным. </div>

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 
    
В случае, если решение может стать еще лучше с некоторыми корректировками </div>
 
<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 
    
В случае, если решение какого-то шага требует значительных изменений. Проект не может быть принят с первого раза, если ревью содержит комментарии, помеченные этим цветом </div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

# Проект на тему: изучение поведения клиентов "Мегалайн"

В проекте я буду работать со следующими данными:

сalls — количество звонков,

minutes — суммарная длительность звонков в минутах,

messages — количество sms-сообщений,

mb_used — израсходованный интернет-трафик в Мб,

is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

Работать буду следующим образом:

1) открою файл и изучу его

2) разделю исходные данные на обучающую, валидационную и тестовую выборки

3) исследую качество разных моделей, меняя гиперпараметры

4) проверю качество модели на тестовой выборке

Приступим!

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error

## Откройте и изучите файл

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')
display(df)
df.info()

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Как и указано в описании проекта, при первичном осмотре файла каких-либо неожиданностей не обнаружено. Продолжим работу без предобработки.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Хороший первичный анализ данных

## Разбейте данные на выборки

In [3]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']
features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.2, random_state=12345) 
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.2, random_state=12345)
print(len(features_test) / len(features))
print(len(features_valid) / len(features))
print(len(features_train) / len(features))
display(features_test)
display(features_valid)
display(features_train)

0.2000622277535781
0.2000622277535781
0.7999377722464219


calls  minutes  messages   mb_used
1415   82.0   507.89      88.0  17543.37
916    50.0   375.91      35.0  12388.40
1670   83.0   540.49      41.0   9127.74
686    79.0   562.99      19.0  25508.19
2951   78.0   531.29      20.0   9217.25
...     ...      ...       ...       ...
2061   66.0   478.48       0.0  16962.58
1510   40.0   334.24      91.0  11304.14
2215   62.0   436.68      52.0  12311.24
664   117.0   739.27     124.0  22818.56
1196   66.0   461.84      85.0  14264.14

[643 rows x 4 columns]

calls  minutes  messages   mb_used
1415   82.0   507.89      88.0  17543.37
916    50.0   375.91      35.0  12388.40
1670   83.0   540.49      41.0   9127.74
686    79.0   562.99      19.0  25508.19
2951   78.0   531.29      20.0   9217.25
...     ...      ...       ...       ...
2061   66.0   478.48       0.0  16962.58
1510   40.0   334.24      91.0  11304.14
2215   62.0   436.68      52.0  12311.24
664   117.0   739.27     124.0  22818.56
1196   66.0   461.84      85.0  14264.14

[643 rows x 4 columns]

calls  minutes  messages   mb_used
348    77.0   522.65      14.0  19462.45
1172  144.0  1031.79      99.0  40174.34
2315   48.0   341.65       0.0  19000.02
954    52.0   383.90      40.0  24574.82
2918   77.0   445.00     100.0  10518.28
...     ...      ...       ...       ...
2817   12.0    86.62      22.0  36628.85
546    65.0   458.46       0.0  15214.25
382   144.0   906.18       0.0  25002.44
2177   38.0   301.27      37.0  28914.24
482   160.0  1084.05      95.0  36843.92

[2571 rows x 4 columns]

Из урока о валидационной выборке разбиваю датафрейм на 3 части в отношении 3:1:1. Проверка на правильность разделения пройдена. 

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Выборки разделены верно, здорово, что проверены их размеры

## Исследуйте модели

Рассмотрим модель DecisionTreeClassifier.

Проверим гиперпараметр max_depth перебором.

In [4]:
for depth in range(1, 10):
    model = model = DecisionTreeClassifier(random_state=12345, max_depth=depth)

    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)

    print("max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1 : 0.7480559875583204
max_depth = 2 : 0.7807153965785381
max_depth = 3 : 0.7838258164852255
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7853810264385692
max_depth = 6 : 0.7822706065318819
max_depth = 7 : 0.7791601866251944
max_depth = 8 : 0.7729393468118196
max_depth = 9 : 0.7791601866251944


Видно, что лучший результат был достигнут при гиперпараметре max_depth равном 5.

Проверим гиперпараметр min_samples_leaf тоже перебором.

In [5]:
for samples in range(1, 100): 
    model = model = DecisionTreeClassifier(random_state=12345, max_depth=5, min_samples_leaf=samples)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    print("min_samples_leaf =", samples, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

min_samples_leaf = 1 : 0.7853810264385692
min_samples_leaf = 2 : 0.7822706065318819
min_samples_leaf = 3 : 0.7822706065318819
min_samples_leaf = 4 : 0.7822706065318819
min_samples_leaf = 5 : 0.7838258164852255
min_samples_leaf = 6 : 0.7947122861586314
min_samples_leaf = 7 : 0.7947122861586314
min_samples_leaf = 8 : 0.7947122861586314
min_samples_leaf = 9 : 0.7900466562986003
min_samples_leaf = 10 : 0.7900466562986003
min_samples_leaf = 11 : 0.7916018662519441
min_samples_leaf = 12 : 0.7916018662519441
min_samples_leaf = 13 : 0.7916018662519441
min_samples_leaf = 14 : 0.7916018662519441
min_samples_leaf = 15 : 0.7916018662519441
min_samples_leaf = 16 : 0.7916018662519441
min_samples_leaf = 17 : 0.7916018662519441
min_samples_leaf = 18 : 0.7916018662519441
min_samples_leaf = 19 : 0.7916018662519441
min_samples_leaf = 20 : 0.7916018662519441
min_samples_leaf = 21 : 0.7916018662519441
min_samples_leaf = 22 : 0.7916018662519441
min_samples_leaf = 23 : 0.7916018662519441
min_samples_leaf = 2

Лучший результат достигнут при значении min_samples_leaf равном 6-8. Пусть будет 6.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

Есть перебор гиперпараметров, верный процесс обучения, но для этого нет смысла создавать новые ячейки, можно было это сделать в одной с помощью подциклов

Посмотрю на модель RandomForestClassifier. При помощи вложенного цикла найду лучшую комбинацию для гиперпараметров max_depth и n_estimators.

In [6]:
max_score = -1
for depth in range(1, 10):
    for est in range(1, 10):
        model = model = RandomForestClassifier(random_state=12345, max_depth=depth, n_estimators=est)

        model.fit(features_train, target_train)
        predictions_valid = model.predict(features_valid)
        score = accuracy_score(target_valid, predictions_valid)
        if score > max_score: 
            print("max_depth =", depth, "n_estimators = ", est, ": ", end='')
            print(score)
            max_score = score

max_depth = 1 n_estimators =  1 : 0.7620528771384136
max_depth = 2 n_estimators =  1 : 0.7869362363919129
max_depth = 2 n_estimators =  4 : 0.7916018662519441
max_depth = 3 n_estimators =  1 : 0.7931570762052877
max_depth = 5 n_estimators =  4 : 0.7978227060653188
max_depth = 6 n_estimators =  7 : 0.7993779160186625
max_depth = 6 n_estimators =  8 : 0.8009331259720062
max_depth = 7 n_estimators =  6 : 0.80248833592535
max_depth = 7 n_estimators =  7 : 0.8087091757387247


Результат показал, что стоит задать гиперпараметры таким образом: max_depth = 7 n_estimators =  7.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

У леса можно побольше гиперпараметров рассмотреть

Посмотрим на модель LogisticRegression(). В ней можно перебрать количество max_iter.

In [7]:
for miter in range(100, 500, 100): 
    model = LogisticRegression(max_iter = miter)
    model.fit(features_train, target_train)
    print(accuracy_score(target_valid, predictions_valid))

0.8040435458786936
0.8040435458786936
0.8040435458786936
0.8040435458786936


Значение на результат не влияет, оставим 100 как и в default.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Ко всему пукнту </b> 
    
Лучше не выводить промежуточные результаты, это громоздко. Если хочешь посмотреть изменение метрики в зависимости от значения перебираемого параметра, то лучше построить график

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Ко всему пукнту </b> 
    
Для более красивого оформления кода можно использовать f-strings. Подробнее про них почитать можешь здесь или на любом другом сайте: https://python-scripts.com/f-strings

Исходя из исследования 3-х моделей, я выбираю RandomForestClassifier как самую точную. Проверю её на тестовой выборке.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

Accuracy - это доля правильных ответов, а не точность. Точность - другая метрика классификации, precision, поэтому про "точную" модель на основе выборки accuracy лучше не говорить

## Проверьте модель на тестовой выборке

In [11]:
model = model = RandomForestClassifier(random_state=12345, max_depth=7, n_estimators=7)
model.fit(features_train, target_train)
predictions_test = model.predict(features_test)
print(accuracy_score(target_test, predictions_test))

0.8087091757387247


Результат accuracy 0.8087 достигнут.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>

Модель проверена на тесте

## (бонус) Проверьте модели на адекватность

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

Можно посмотреть на баланс классов: например, если у нас 90 процентов таргета это "1", а остальные 10 - "0", то мы можем тупо выдавать каждый раз цифру 1 и получать метрику 90 %. Любая модель, которая будет выдавать качество ниже этого - плохая/неадекватная
    
dummy classifier с strategy='most_frequent' работает так же: смотрит на самый часто встречающийся класс и выдает только его (только одно значение - константу, поэтому это константная модель)

# Вывод:

В результате выполненной работы наиболее точной, как и ожидалось, показала себя модель RandomForestClassifier, которая достигла результата accuracy 0.8087.


h <div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

Хороший проект, тобою проделана большая работа, ты молодец!

Сперва похвалю за наличие везде промежуточных выводов, комментариев. Хороший код, показываешь навыки владения sklearn. Все требуемые пункты в задании в целом выполнены. Оставил пару желтых комментариев, думаю, это сможет улучшить твой проект. Также можно добавить визуализацию в проект: посмотреть на графиках, как влияет перебор параметров на результат нашей метрики

В этом проекте большое внимание уделяется подбору гиперпараметров, но это можно автоматизировать! Советую почитать на досуге про GridsearchCV и Optuna, будет полезно
            
Из курсов и дополнительной информации советую Yandex ML Handbook (https://academy.yandex.ru/handbook/ml) и курсы Евгения Соколова на ФКН (Машинное обучение - 1 с математикой https://github.com/esokolov/ml-course-hse и Основы машинного обучения без глубокой математики https://github.com/hse-ds/iad-intro-ds) - там можешь получить дополнительную информацию, более глубокую, возможно. http://wiki.cs.hse.ru/Заглавная_страница - тут можешь глянуть записи семинаров и лекций и другие материалы соответствующих курсов. 
        
Готов принять проект и так, потому что не было ошибок. Можешь отправлять проект без доработок, но если хочешь - можешь исправить желтые комментарии/задать свои вопросы по моим комментариям, проекту, ML в целом
    
Удачи!

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
